File: scripting/engine/Playground/NativeLookupTable.swift
internal enum NativeLookupTable {
    internal nonisolated(unsafe) static var table: [String: UnsafeRawPointer] = [:]

    internal nonisolated static func getFunctionPointer<T>(by name: String) -> T {
        let ptr = unsafeBitCast(Self.table[name], to: T.self)

        return ptr
    }

    internal nonisolated static func addEntry(key: String, ptr: UnsafeRawPointer) {
        Self.table[key] = ptr
    }
}

File: engine/managed/PlaygroundAssembly/Playground.Core/NativeLookupTable.cs
using System.Text;

namespace Playground.Core;

internal static unsafe class NativeLookupTable
{
    private static readonly Dictionary<string, IntPtr> Table = new();

    internal static void* GetFunctionPointer(string name)
    {
        return (void*)Table[name];
    }

    internal static void AddEntry(byte* key, int length, IntPtr ptr)
    {
        var strName = Encoding.UTF8.GetString(key, length);
        Table[strName] = ptr;
    }
}
